### ISSUE TITLE
Implement read controls (RFC4527)

### ISSUE THREAD

**Issue by Elizafox** [cannatag/ldap3#273]
Hello,

It would be nice if ldap3 implemented [RFC4527](https://tools.ietf.org/html/rfc4527) to make the auto-increment extension useful for implementing UID's (without it, the update is potentially non-atomic), or had some constant which implemented it that could be passed via `controls`.

I can't figure out how to make the existing controls parameter to `Connection.modify` accept an `AttributeSelection` object from `ldap3.protocol.rfc4511` (if that is the correct way to do it; the RFC seems to say that the parameter needs to be formed as an `AttributeSelection`, unless I am misinterpreting it).

**Comment by cannatag** [comment]
Hi, it should be quite simple to build the control request. Also decoding the control response should be easy because it includes a full search response. What LDAP server accept this controls? Are you able to perform some testing from dev code?

Bye,
Giovanni

**Comment by Elizafox** [comment]
Hey,

I think I cracked it, but it's definitely not what I would call "ideal":

```from ldap3 import Server, Connection, NO_ATTRIBUTES, ALL_ATTRIBUTES, MODIFY_REPLACE
from ldap3.operation.search import build_attribute_selection
from ldap3.protocol.controls import build_control

def read_control(oid, attributes, criticality=False, schema=None):
    if not attributes:
        attributes = [NO_ATTRIBUTES]
    elif attributes == ALL_ATTRIBUTES:
        attributes = [ALL_ATTRIBUTES]

    value = build_attribute_selection(attributes, schema)
    return build_control(oid, criticality, value)

def pre_read_control(attributes, criticality=False, schema=None):
    return read_control('1.3.6.1.1.13.1', attributes, criticality, schema)

def post_read_control(attributes, criticality=False, schema=None):
    return read_control('1.3.6.1.1.13.2', attributes, criticality, schema)

server = Server('localhost')
conn = Connection(server, 'cn=admin,dc=glpgs,dc=io', 'ChangeTh1sC4t', auto_bind=True)

conn.modify('cn=uidnext,dc=glpgs,dc=io',
            {'uidNumber' : [(MODIFY_REPLACE, [10002])]},
            controls=[pre_read_control(["uidNumber"], True)])

print(conn.result)
```

conn.result gives me the following:

`{'dn': '', 'type': 'modifyResponse', 'description': 'success', 'message': '', 'result': 0, 'controls': {'1.3.6.1.1.13.1': {'criticality': False, 'value': b'd3\x04\x19cn=uidnext,dc=glpgs,dc=io0\x160\x14\x04\tuidNumber1\x07\x04\x0510002', 'description': ('1.3.6.1.1.13.1', 'CONTROL', 'LDAP Pre-read', 'RFC4527')}}, 'referrals': None}`

**Comment by Elizafox** [comment]
Hey,

So I figured out how to parse the result. I will be sending some patches to properly decode the responses soon without needing to hack about.

**Comment by Elizafox** [comment]
Hey,

So I'm writing a patch to parse the value returned into something usable. Do you want the resulting values returned from the server placed in the `Connection.entries` object, or should it just be returned in the `Connection.result` object?

**Comment by Elizafox** [comment]
I removed my previous comment and decided to just implement it. My implementation is here: https://github.com/cannatag/ldap3/pull/274

**Comment by cannatag** [comment]
Will be released in version 2.2

**Comment by cannatag** [comment]
included in v2.2.0
